src/main/java/com/snowflake/kafka/connector/internal/streaming/schemaevolution/iceberg/IcebergFieldNode.java

```diff
@@ -61,26 +68,145 @@ StringBuilder buildQuery(StringBuilder sb, String parentType) {
     return sb;
   }
 
-  private StringBuilder appendNameAndType(StringBuilder sb) {
+  /**
+   * Method does not modify, delete any existing nodes and its types, names. It is meant only to add
+   * new children.
+   */
+  void merge(IcebergFieldNode modifiedNode) {
+    modifiedNode.children.forEach(
+        (key, node) -> {
+          IcebergFieldNode thisChild = this.children.get(key);
+          if (thisChild == null) {
+            this.children.put(key, node);
+          } else {
+            thisChild.merge(node);
+          }
+        });
+  }
+
+  private LinkedHashMap<String, IcebergFieldNode> produceChildren(JsonNode recordNode) {
+    if (recordNode.isNull()) {
+      return new LinkedHashMap<>();
+    }
+    if (recordNode.isArray()) {
+      ArrayNode arrayNode = (ArrayNode) recordNode;
+      return produceChildrenFromArray(arrayNode);
+    }
+    if (recordNode.isObject()) {
+      ObjectNode objectNode = (ObjectNode) recordNode;
+      return produceChildrenFromObject(objectNode);
+    }
+    return new LinkedHashMap<>();
+  }
+
+  private LinkedHashMap<String, IcebergFieldNode> produceChildrenFromArray(ArrayNode arrayNode) {
+    JsonNode arrayElement = arrayNode.get(0);
+    // VARCHAR is set for an empty array: [] -> ARRAY(VARCHAR)
+    if (arrayElement == null) {
+      LinkedHashMap<String, IcebergFieldNode> child = new LinkedHashMap<>();
+      child.put("element", new IcebergFieldNode("element", "VARCHAR(16777216)"));
+      return child;
+    }
+    LinkedHashMap<String, IcebergFieldNode> child = new LinkedHashMap<>();
+    child.put("element", new IcebergFieldNode("element", arrayElement));
+    return child;
+  }
+
+  private LinkedHashMap<String, IcebergFieldNode> produceChildrenFromObject(ObjectNode objectNode) {
+    return objectNode.properties().stream()
+        .collect(
+            Collectors.toMap(
+                Map.Entry::getKey,
+                stringJsonNodeEntry ->
+                    new IcebergFieldNode(
+                        stringJsonNodeEntry.getKey(), stringJsonNodeEntry.getValue()),
+                (v1, v2) -> {
+                  throw new IllegalArgumentException("Two same keys: " + v1);
+                },
+                LinkedHashMap::new));
+  }
+
+  private LinkedHashMap<String, IcebergFieldNode> produceChildren(Schema connectSchema) {
+    Schema.Type type = connectSchema.type();
+    if (connectSchema.type() == Schema.Type.STRUCT) {
+      return produceChildrenFromStruct(connectSchema);
+    }
+    if (connectSchema.type() == Schema.Type.MAP) {
+      return produceChildrenFromMap(connectSchema);
+    }
+    if (connectSchema.type() == Schema.Type.ARRAY) {
+      return produceChildrenForArray(connectSchema);
+    } else { // isPrimitive == true
+      return new LinkedHashMap<>();
+    }
```

Comment:
i would skip the else statement for readability
```suggestion
    } 
    // primitive types don't have children
    return new LinkedHashMap<>();
    
```